services/scanner/agentpool/poolagent/agent.go

```diff
@@ -219,46 +227,90 @@ func (agent *Agent) SetClient(agentClient clients.AgentClient) {
 // from request channels.
 func (agent *Agent) StartProcessing() {
 	agent.initWait.Add(1)
-	go agent.initialize()
 
+	go func() {
+		err := agent.initWorker(agent.ctx)
+		if err != nil {
+			log.WithError(err).Warn("failed to initialize bot")
+		}
+	}()
 	go agent.processTransactions()
 	go agent.processBlocks()
 	go agent.processCombinationAlerts()
 }
 
-func (agent *Agent) initialize() {
-	defer agent.initWait.Done()
+func (agent *Agent) initWorker(ctx context.Context) error {
+	// first try without waiting for the ticker
+	initCtx, cancel := context.WithTimeout(ctx, DefaultAgentInitializeTimeout)
+	if err := agent.initialize(initCtx); err == nil {
+		// 	break if successfully initialized
+		cancel()
+		return nil
+	}
+	cancel()
+
+
+	// start retrying
+	initRetryInterval := time.Minute * 10
+	t := time.NewTicker(initRetryInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+			initCtx, cancel := context.WithTimeout(agent.ctx, DefaultAgentInitializeTimeout)
+			if err := agent.initialize(initCtx); err == nil {
+				// 	break if successfully initialized
+				cancel()
+				return nil
+			}
+			cancel()
+		}
+	}
+}
 
-	logger := log.WithFields(log.Fields{
-		"agent": agent.config.ID,
-	})
+func (agent *Agent) initialize(ctx context.Context) error {
+	logger := log.WithFields(
+		log.Fields{
+			"agent": agent.config.ID,
+		},
+	)
 
-	ctx, cancel := context.WithTimeout(agent.ctx, DefaultAgentInitializeTimeout)
-	defer cancel()
-	initializeResponse, err := agent.client.Initialize(ctx, &protocol.InitializeRequest{
-		AgentId:   agent.config.ID,
-		ProxyHost: config.DockerJSONRPCProxyContainerName,
-	})
+	initializeResponse, err := agent.client.Initialize(
+		ctx, &protocol.InitializeRequest{
+			AgentId:   agent.config.ID,
+			ProxyHost: config.DockerJSONRPCProxyContainerName,
+		},
+	)
 
 	if status.Code(err) == codes.Unimplemented {
 		logger.WithError(err).Info("initialize() method not implemented in bot - safe to ignore")
-		return
+		return err
 	}
+
 	if err != nil {
 		logger.WithError(err).Warn("bot initialization failed")
-		return
+		return err
 	}
 
-	if err := validateInitializeResponse(initializeResponse); err != nil{
+	if err := validateInitializeResponse(initializeResponse); err != nil {
 		logger.WithError(err).Warn("bot initialization validation failed")
-		return
+		return err
 	}
 
+	// pass new alert subscriptions to pool
 	if initializeResponse != nil {
 		agent.SetAlertConfig(initializeResponse.AlertConfig)
+		for _, subscription := range initializeResponse.AlertConfig.Subscriptions {
+			agent.newSubscriptions <- messaging.CombinerBotSubscription{Subscription: subscription}
```

Comment:
We're passing the subscriptions to another goroutine via an unbuffered channel. It is technically not much different than making the `Publish()` call here synchronously instead. I suggest removing `handleAlertSubscriptions()` and the new subscription channel in favor of this.

Reply:
In terms of async behavior, you're right. But the general approach in the agent pool was that the `AgentPool` makes the communications, so I didn't want to break that design.

Reply:
Got it - It feels like the `AgentPool` doesn't add anything special to this communication and we have `Publish()`es in a lot of arbitrary places already.